### PR TITLE
Adds simple wrapper for a cancellable context

### PR DESCRIPTION
**Pull request by markylaing**
When a resource such as a database must advertise when it becomes ready, we can use a cancellable `context.Context`. This is useful because subsequent calls to the associated `CancelFunc` will not cause the goroutine to panic (whereas calling `close()` on a channel will). However this means that the resource must keep a reference to both the cancellable context and it's `CancelFunc` which is less desirable.

This change adds a simple wrapper to a cancellable context such that it can be used like:
```golang
canceller := cancel.New(ctx)
defer canceller.Close()
```

Additionally, the existing canceller was renamed and moved to `shared/cancel/http.go`.

Based on chat here https://github.com/canonical/lxd-cloud/pull/114#discussion_r860759810

### CODE REVIEW

File: shared/cancel/canceller.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package cancel
+
+import "context"
+
+// Canceller is a simple wrapper for a cancellable context which makes the associated context.CancelFunc more easily
+// accessible.
+type Canceller struct {
+	context.Context
+	Close func()
```
Comment (tomponline) [inline shared/cancel/canceller.go:9]: Shall we call this `Cancel` actually, as that keeps it inline with the `HTTPRequestCanceller.Cancel()` function and the context.Context convention of using the `cancel` function var name.

**Review by tomponline — CHANGES_REQUESTED** [review]
Thanks! Apart from renaming `Close` I think this is good to go.

File: shared/cancel/canceller.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+}
+
+// New returns a new canceller with the parent context.
+func New(ctx context.Context) Canceller {
```
Comment (tomponline) [inline shared/cancel/canceller.go:13]: This should return a *Canceller

**Comment by markylaing** [comment]
@tomponline should be good to go

**Review by tomponline — APPROVED** [review]

File: shared/cancel/canceller.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package cancel
+
+import "context"
+
+// Canceller is a simple wrapper for a cancellable context which makes the associated context.CancelFunc more easily
+// accessible.
+type Canceller struct {
+	context.Context
+	Cancel func()
```
Comment (tomponline) [inline shared/cancel/canceller.go:9]: @markylaing oh can we make the type of this field `context.CancelFunc` for clarity.

File: shared/cancel/canceller.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package cancel
+
+import "context"
+
+// Canceller is a simple wrapper for a cancellable context which makes the associated context.CancelFunc more easily
+// accessible.
+type Canceller struct {
+	context.Context
+	CancelFunc func()
```
Comment (tomponline) [inline shared/cancel/canceller.go:9]: @markylaing sorry you misunderstood me, I meant like this `Cancel context.CancelFunc`

Comment (markylaing) [reply shared/cancel/canceller.go:9]: Ohhh ok sure

**Pull request merged**